convert2rhel/toolopts.py

```diff
@@ -279,9 +288,27 @@ def _process_cli_options(self):
             tool_opts.password = parsed_opts.password
 
         if parsed_opts.password_from_file:
+            loggerinst.warning("Deprecated. Use -c | --config-file instead")
             tool_opts.password_file = parsed_opts.password_from_file
             tool_opts.password = utils.get_file_content(parsed_opts.password_from_file)
 
+        # Processing the configuration file
+        if parsed_opts.config_file or (not parsed_opts.password or not parsed_opts.activationkey):
```

Comment:
If the second part of condition would be `(not parsed_opts.password and not parsed_opts.activationkey)`  the difference will be in [this variant of test](https://github.com/oamg/convert2rhel/pull/451/files#diff-5537d9a47132079b0eff4fb09b1247db109b6322ee99aad12628df6772136346R166) with output value equal to `{"password": None, "activation-key": "activation-key"}`. 

I think it's fine, the activation key has precedence. And maybe sometime in the future the convert2rhel could try to use it if the activation key fails?

Reply:
I think it would be better to always set `tool_opts.config_file` here since we might have other, non-secret, values in the config file at a later date.

In other projects, I parse the config file before I look into any of the other command line arguments and then, when I merge that with the command line arguments.

Our password and activation parameters are special, though, as they have an interrelation.  I would set the tool_opts for those in one place instead of setting the tool_opts from the cli first, and then setting it from the config file in a later section.  So maybe something like this:

``` python
parsed_opts, _ = self._parser.parse_args()
tool_opts.config_file = parsed_opts.config_file

# Notes:
# options_from_config_files reads from all of the config files
# It skips the passed in config file if it is None (no config file specified on the CLI)
# It raises an exception if the passed in config file is specified but does not exist/is unreadable
# It ignores if one of the hardcoded config file locations does not exist.  It may be wise to warn if it is unreadable.
# It merges the values from the hardcoded and passed in locations.  The passed in location takes precedence followed by ~/.convert2rhel.ini, and then /etc/convert2rhel.ini if neither of the other two have a value.

config_opts = options_from_config_files(parsed_opts.config_file)

# General config options (which don't have cornercases like password + activation key)
# Note, we don't have any of these in the config file yet.
tool_opts.server_url = config_opts.copy()
if parsed_opts.debugl:
    tool_opts.debug = parsed_opts.debug
# [..]

if parsed_opts.password or parsed_opts.activationkey:
    if parsed_opts.password and parsed_opts.activationkey:
        loggerinst.warning("Set only one of password or activation key")
    tool_opts.password = tool_opts.activationkey = None
    if parsed_opts.activationkey:
        tool_opts.activationkey = parsed_opts.activationkey
    if parsed_opts.password:
        tool_opts.password = parsed_opts.password
```

Reply:
@abadger by this suggestion, do you mean to say we hard-code the reading of all entries in the ini file? Would it not be better for us to iterate through all entries in the ini and see if it matches our keys in tool_opts?
```py
# General config options (which don't have cornercases like password + activation key)
# Note, we don't have any of these in the config file yet.
tool_opts.server_url = config_opts.copy()
```

But yes I agree that we should 
1. Parse entries from .ini files in terms of priority
2. Overwrite values set by .ini files if CLI options are set

Reply:
@abadger @SpyTec Maybe the idea with iterating through the entries in file would be better. I'll try it that way - I was focused just on doing it in one way, this variant didn't come to my mind :smile: 

But anyway, parsing the .ini file at first time and then overwrite the values by values from CLI simplified the part which I called 'Processing the configuration file'.

@r0x0d I think you are half right: When are provided both (password and activation-key) it will be caught by another condition which mentioned @abadger and it would say 'loggerinst.warning("Set only one of password or activation key")'. 

You wrote that we are ignoring password or activation-key when we detect config file. That's not right, because the CLI options have precedence over the options from config file.